graphql/schemabuilder/reflect.go

```diff
@@ -315,63 +325,74 @@ func wrapPtrParser(inner *argParser) *argParser {
 }
 
 func (sb *schemaBuilder) makeStructParser(typ reflect.Type) (*argParser, graphql.Type, error) {
-	fields := make(map[string]argField)
-	argType := &graphql.InputObject{
-		Name:        typ.Name(),
-		InputFields: make(map[string]graphql.Type),
-	}
-	if argType.Name != "" {
-		argType.Name += "_InputObject"
-	}
-
-	if typ.Kind() != reflect.Struct {
-		return nil, nil, fmt.Errorf("expected struct but received type %s", typ.Name())
-	}
-
-	for i := 0; i < typ.NumField(); i++ {
-		field := typ.Field(i)
-		if field.PkgPath != "" && !field.Anonymous {
-			continue
-		}
-		if field.Anonymous {
-			return nil, nil, fmt.Errorf("bad arg type %s: anonymous fields not supported", typ)
+	var fields map[string]argField
+	var argType *graphql.InputObject
+	cached, ok := typeCache[typ]
+	if ok {
+		fields = cached.fields
+		argType = cached.argType
+	} else {
+		fields = make(map[string]argField)
+		argType = &graphql.InputObject{
+			Name:        typ.Name(),
+			InputFields: make(map[string]graphql.Type),
 		}
-		tags := strings.Split(field.Tag.Get("graphql"), ",")
-		var name string
-		if len(tags) > 0 {
-			name = tags[0]
+		if argType.Name != "" {
+			argType.Name += "_InputObject"
 		}
-		if name == "" {
-			name = makeGraphql(field.Name)
-		}
-		if name == "-" {
-			continue
+
+		if typ.Kind() != reflect.Struct {
+			return nil, nil, fmt.Errorf("expected struct but received type %s", typ.Name())
 		}
 
-		var key bool
+		// Cache type information ahead of time to catch self-reference
+		typeCache[typ] = cachedType{argType, fields}
```

Comment:
could you add a test for this this case?

Reply:
could you also document in the commit message what the previous behavior used to be?

Reply:
Hey 👋, sure, I'll work on it some more next week-ish.